numpy/linalg/linalg.py

```diff
@@ -2745,3 +2752,99 @@ def _multi_dot(arrays, order, i, j):
+_noop_get_num_threads = lambda *args, **kwargs: 1
+
+if cdll is None:
+    set_num_threads = set_module('numpy.linalg')(_noop_set_num_threads)
+    get_num_threads = set_module('numpy.linalg')(_noop_get_num_threads)
+else:
+    _cached_threads_funcs = [None, None]
+
+    @set_module('numpy.linalg')
+    def set_num_threads(nthreads):
+        """
+        Specify the number of OpenMP threads to use in further calls to BLAS.
+        MKL and OpenBLAS are currently supported.
+
+        Notes
+        -----
+        Changing the number of threads is not a light task, this should
+        be performed only if necessary.
+        """
+        _set_num_threads = _cached_threads_funcs[0]
+        if _set_num_threads is None:
+            blas = cdll[lapack_lite.__file__]
+            if blas is None:
+                _set_num_threads = _noop_set_num_threads
+            else:
+                try:
+                    blas.MKL_Set_Num_Threads.argtypes = [c_int]
+                    _set_num_threads = blas.MKL_Set_Num_Threads
+                except AttributeError:
+                    pass
+            if _set_num_threads is None:
+                try:
+                    blas.openblas_set_num_threads.argtypes = [c_int]
+                    _set_num_threads = blas.openblas_set_num_threads
+                except AttributeError:
+                    _set_num_threads = _noop_set_num_threads
+            _cached_threads_funcs[0] = _set_num_threads
+        _set_num_threads(nthreads)
+
+    @set_module('numpy.linalg')
+    def get_num_threads():
+        """
+        Get the number of OpenMP threads used by BLAS.
+        MKL and OpenBLAS are currently supported.
+        """
+        _get_num_threads = _cached_threads_funcs[1]
+        if _get_num_threads is None:
+            blas = cdll[lapack_lite.__file__]
+            if blas is None:
+                _get_num_threads = _noop_get_num_threads
+            else:
+                try:
+                    blas.MKL_Get_Max_Threads.restype = c_int
+                    _get_num_threads = blas.MKL_Get_Max_Threads
+                except AttributeError:
+                    pass
+            if _get_num_threads is None:
+                try:
+                    blas.openblas_get_num_threads.restype = c_int
+                    _get_num_threads = blas.openblas_get_num_threads
+                except AttributeError:
+                    _get_num_threads = _noop_get_num_threads
+            _cached_threads_funcs[1] = _get_num_threads
+        return _get_num_threads()
+
+
+@set_module('numpy.linalg')
+@contextlib.contextmanager
+def num_threads(nthreads):
+    """Context manager for temporarily changing number of threads in BLAS.
+
+    Change number of threads for the scope of the `with` block, and restore the
+    old number of threads at the end. This may be desirable to prevent
+    multithreading in BLAS if this is known to cause performance issues.
+
```

Comment:
it would be quite useful if one of these docstrings (I suggest this one) had a bit of an explanation of how openblas and mkl use threads, what env vars this overrides, etc.

Reply:
How about something like "OpenBLAS and MKL uses a multi-threading model based on OpenMP. By default the number of threads will be the number of CPUs on the machine. On a machine with many cores and many concurrent OpenMP processes this can quickly overallocate resources, causing thread thrashing and overall processing slowdown. The number of threads can be set before running the process via the environment variables OMP_NUM_THREADS (for OpenMP), OPENBLAS_NUM_THREADS (for OpenBLAS) or MKL_NUM_THREADS (for mkl)."

Also maybe mention in the first line that the change is not local but affects the entire process space (influencing any other threads) - "Change number of threads used by the OpenBLAS or MKL subsytem in any threads the process for the scope of the `with` block"

Both of these are suggestions and should be worded better

Reply:
that sounds good, that captures the most essential things.

maybe in addition a note on the different ways this works depending on how OpenBLAS was built (IIRC you can enable/disable OpenMP and set the max number of threads at build time.

Reply:
I think the documentation should probably also mention that the context manager is not thread-safe.